### PR TITLE
Decode and Encode of enum  AND unit

### LINKED ISSUE #25: Handling of Enums
Currently encoding and decoding of ordinary `enum`s like `System.DayOfTheWeek` require own extra coders, e.g.
```fsharp
module Encode =
   let dayofWeek (value: System.DayOfWeek) : JsonValue =
       LanguagePrimitives.EnumToValue (value)
       |>JValue :> JsonValue


module Decode =
    module Helpers =
        let inline isInteger (token: JsonValue) = not(isNull token) && (token.Type = JTokenType.Integer)
        let inline isString (token: JsonValue) = not(isNull token) && token.Type = JTokenType.String
        let inline asString (token: JsonValue): string = token.Value<string>()
    
    let dayofWeek : Decoder<System.DayOfWeek> =
        fun path token ->
            if Helpers.isInteger token then
                // TODO: Is not enough to convert to int directly? Maybe these checks hurt performance?
                let value = token.Value<decimal> ()
                if value >= (decimal System.Int32.MinValue) && value <= (decimal System.Int32.MaxValue) then
                    Ok (LanguagePrimitives.EnumOfValue ((int32) value))
                else
                    (path, BadPrimitiveExtra("an int", token, "Value was either too large or too small for an int")) |> Error
            elif Helpers.isString token then
                match System.Int32.TryParse (Helpers.asString token, NumberStyles.Any, CultureInfo.InvariantCulture) with
                | true, x -> Ok <| LanguagePrimitives.EnumOfValue(x)
                | _ -> (path, BadPrimitive("an int", token)) |> Error
            else
                (path, BadPrimitive("an int", token)) |> Error

```

This works pretty good, however I would love to have generic Enum support, either direclty in Thoth.Json or as extra coder.

This naive approach fail:
```fsharp
module Encode =

   let enumInt (value: 'T when 'T:enum<int>) : JsonValue =
       LanguagePrimitives.EnumToValue (value)
       |>JValue :> JsonValue

module Decode =
    module Helpers =
        let inline isInteger (token: JsonValue) = not(isNull token) && (token.Type = JTokenType.Integer)
        let inline isString (token: JsonValue) = not(isNull token) && token.Type = JTokenType.String
        let inline asString (token: JsonValue): string = token.Value<string>()
    
    let enumInt : Decoder<'T> when 'T:enum<int> =
        fun path token ->
            if Helpers.isInteger token then
                // TODO: Is not enough to convert to int directly? Maybe these checks hurt performance?
                let value = token.Value<decimal> ()
                if value >= (decimal System.Int32.MinValue) && value <= (decimal System.Int32.MaxValue) then
                    Ok (LanguagePrimitives.EnumOfValue ((int32) value))
                else
                    (path, BadPrimitiveExtra("an int", token, "Value was either too large or too small for an int")) |> Error
            elif Helpers.isString token then
                match System.Int32.TryParse (Helpers.asString token, NumberStyles.Any, CultureInfo.InvariantCulture) with
                | true, x -> Ok <| LanguagePrimitives.EnumOfValue(x)
                | _ -> (path, BadPrimitive("an int", token)) |> Error
            else
                (path, BadPrimitive("an int", token)) |> Error
```
While the encoders and decoders work, they cannot be added `withCustom`, and I also failt adding them directly in Thoth.Json.Net.

It worked before in 2.x, maybe this just a special case of #9, where all unknown types case were handled by NewtonSoft.Json.

### PR DESCRIPTION
**Pull request by SCullman**
Fixes #25

### CODE REVIEW

File: src/Encode.fs [hunk 1]
```diff
@@ -189,8 +189,9 @@ module Encode =
     let timespan (value : System.TimeSpan) : JsonValue =
         JValue(value.ToString()) :> JsonValue
 
-    let enum (value) : JsonValue =
-        JValue(value.ToString()) :> JsonValue
+    let enum (value: 'T when 'T:enum<int>) =
```
Comment (SCullman) [inline src/Encode.fs]: enum is not used for Auto Encoding

**Comment by SCullman** [comment]
Tests are in Thoth.Net, tested locally and not run here in CI, as this require a merge in https://github.com/thoth-org/Thoth.Json/pull/25

File: src/Encode.fs (hunk 1 above)
Comment (SCullman) [reply src/Encode.fs]: fixed

**Comment by MangelMaxime** [comment]
Closed in favour of https://github.com/thoth-org/Thoth.Json.Net/pull/27 which is in par with Thoth.Json API.

**Pull request closed without merging**